facade/src/main/java/org/jboss/pnc/facade/providers/ProductMilestoneProviderImpl.java

```diff
@@ -359,4 +386,187 @@ private List<Integer> getDependentMilestoneIds(CriteriaBuilder cb, String id) {
+        Join<Artifact, BuildRecord> builtArtifacts = artifacts.join(Artifact_.buildRecord);
+        Join<BuildRecord, org.jboss.pnc.model.ProductMilestone> builtArtifactsMilestones = builtArtifacts
+                .join(BuildRecord_.productMilestone);
+
+        query.where(cb.equal(builtArtifactsMilestones.get(ProductMilestone_.id), getEntityIdFromRestId(id)));
+        query.select(cb.count(artifacts.get(Artifact_.id)));
+
+        return em.createQuery(query).getSingleResult();
+    }
+
+    private Long getDeliveredArtifactsBuiltInThisMilestone(CriteriaBuilder cb, String id) {
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<Artifact> artifacts = query.from(Artifact.class);
+        SetJoin<Artifact, org.jboss.pnc.model.ProductMilestone> deliveredArtifacts = artifacts
+                .join(Artifact_.deliveredInProductMilestones);
+        Join<Artifact, BuildRecord> builtArtifacts = artifacts.join(Artifact_.buildRecord);
+
+        Integer productMilestoneId = getEntityIdFromRestId(id);
+        query.where(
+                cb.equal(
+                        builtArtifacts.get(BuildRecord_.productMilestone).get(ProductMilestone_.id),
+                        productMilestoneId),
+                cb.equal(deliveredArtifacts.get(ProductMilestone_.id), productMilestoneId));
+        query.select(cb.count(artifacts.get(Artifact_.id)));
+
+        return em.createQuery(query).getSingleResult();
+    }
+
+    private Long getDeliveredArtifactsBuiltInOtherMilestones(CriteriaBuilder cb, String id) {
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<Artifact> artifacts = query.from(Artifact.class);
+        SetJoin<Artifact, org.jboss.pnc.model.ProductMilestone> deliveredArtifacts = artifacts
+                .join(Artifact_.deliveredInProductMilestones);
+        Join<Artifact, BuildRecord> builtArtifacts = artifacts.join(Artifact_.buildRecord);
+        Join<BuildRecord, org.jboss.pnc.model.ProductMilestone> milestoneBuiltArtifacts = builtArtifacts
+                .join(BuildRecord_.productMilestone);
+        Join<org.jboss.pnc.model.ProductMilestone, ProductVersion> versionBuiltArtifacts = milestoneBuiltArtifacts
+                .join(ProductMilestone_.productVersion);
+        Join<ProductVersion, Product> productBuiltArtifacts = versionBuiltArtifacts.join(ProductVersion_.product);
+
+        Integer productMilestoneId = getEntityIdFromRestId(id);
+        Integer productId = getProductIdByItsMilestone(cb, productMilestoneId);
+        query.where(
+                cb.equal(deliveredArtifacts.get(ProductMilestone_.id), productMilestoneId),
+                cb.equal(productBuiltArtifacts.get(Product_.id), productId),
+                cb.notEqual(milestoneBuiltArtifacts.get(ProductMilestone_.id), productMilestoneId));
+        query.select(cb.count(artifacts.get(Artifact_.id)));
+
+        return em.createQuery(query).getSingleResult();
```

Comment:
Maybe something like this? Also for other cases
```suggestion
        CriteriaQuery<Long> query = cb.createQuery(Long.class);

        Root<org.jboss.pnc.model.ProductMilestone> milestone = query.from(org.jboss.pnc.model.ProductMilestone.class);
        SetJoin<org.jboss.pnc.model.ProductMilestone, Artifact> deliveredArtifacts = milestone.
                join(ProductMilestone_.deliveredArtifacts);

        Path<org.jboss.pnc.model.ProductMilestone> artifactsMilestone = deliveredArtifacts.
                join(Artifact_.buildRecord).get(BuildRecord_.productMilestone);

        query.where(
                cb.equal(milestone.get(ProductMilestone_.id), id),
                cb.equal(artifactsMilestone.get(ProductMilestone_.productVersion).get(ProductVersion_.product),
                        milestone.get(ProductMilestone_.productVersion).get(ProductVersion_.product)),
                cb.notEqual(artifactsMilestone.get(ProductMilestone_.id), id));
        query.select(cb.count(deliveredArtifacts.get(Artifact_.id)));

        return em.createQuery(query).getSingleResult();
```